main/src/ca/uwaterloo/flix/language/phase/MonomorphEnums.scala

```diff
@@ -0,0 +1,698 @@
+      val e = visitExp(exp)
+      val t = visitType(tpe)
+      val p = visitType(pur)
+      Expression.Force(e, t, p, loc)
+  }
+
+  /**
+    * Returns a pattern with specialized enums in its type and no aliases.
+    */
+  private def visitPat(pat: LoweredAst.Pattern)(implicit ctx: Context, root: LoweredAst.Root, flix: Flix): LoweredAst.Pattern = pat match {
+    case Pattern.Wild(tpe, loc) =>
+      val t = visitType(tpe)
+      Pattern.Wild(t, loc)
+    case Pattern.Var(sym, tpe, loc) =>
+      val t = visitType(tpe)
+      Pattern.Var(sym, t, loc)
+    case Pattern.Cst(cst, tpe, loc) =>
+      val t = visitType(tpe)
+      Pattern.Cst(cst, t, loc)
+    case Pattern.Tag(sym, tagPat, tpe, loc) =>
+      //
+      // Specialize the enum
+      //
+      val freshCaseSym = specializeCaseSymUse(sym, tpe.typeArguments, tpe.loc)
+      val tp = visitPat(tagPat)
+      val t = visitType(tpe)
+      Pattern.Tag(freshCaseSym, tp, t, loc)
+    case Pattern.Tuple(elms, tpe, loc) =>
+      val es = elms.map(visitPat)
+      val t = visitType(tpe)
+      Pattern.Tuple(es, t, loc)
+  }
+
+  /**
+    * Returns a type with specialized enums and no aliases.
+    */
+  private def visitType(tpe: Type)(implicit ctx: Context, root: LoweredAst.Root, flix: Flix): Type = {
+    def visitInner(tpe0: Type): Type = tpe0.baseType match {
+      case Type.Cst(TypeConstructor.Enum(sym, _), loc) => // enum
+        // args cannot be visited yet, because specialization works on
+        // non-specialized enums.
+        val args = tpe0.typeArguments
+        val freshSym = specializeEnum(sym, args, loc)
+        Type.mkEnum(freshSym, Nil, loc)
+      case _ => tpe0 match { // non-enum
+        case Type.Cst(tc, loc) => Type.Cst(tc, loc)
+        case Type.Apply(tpe1, tpe2, loc) => Type.Apply(visitInner(tpe1), visitInner(tpe2), loc)
+        case Type.Var(sym, loc) => throw InternalCompilerException(s"Unexpected type var: '$sym'", loc)
+        case Type.Alias(cst, _, _, loc) => throw InternalCompilerException(s"Unexpected type alias: '${cst.sym}'", loc)
+        case Type.AssocType(cst, _, _, loc) => throw InternalCompilerException(s"Unexpected associated type: '${cst.sym}'", loc)
+      }
+    }
+    // It is important that eraseAliases happens BEFORE enum specialization
+    visitInner(eraseAliases(tpe))
+  }
+
+  /**
+    * Returns a formal param with specialized enums in its type and no aliases.
+    */
+  private def visitFormalParam(p: LoweredAst.FormalParam)(implicit ctx: Context, root: LoweredAst.Root, flix: Flix): LoweredAst.FormalParam = {
+    val LoweredAst.FormalParam(sym, mod, tpe, src, loc) = p
+    val t = visitType(tpe)
+    LoweredAst.FormalParam(sym, mod, t, src, loc)
+  }
+
+  /**
+    * Returns a scheme with specialized enums in its base and no aliases.
+    */
+  private def visitScheme(sc: Scheme)(implicit ctx: Context, root: LoweredAst.Root, flix: Flix): Scheme = {
+    val Scheme(quantifiers, tconstrs, econstrs, base) = sc
```

Comment:
Not visiting the other stuff? In theory it does not matter since tconstrs and econstrs shouldn't be used anymore. But maybe worth mentioning.